### PR TITLE
Fixed sort-by not sorting Resources as expected

### LINKED ISSUE #1019: sort-by not sorting as expected
**What happened**: When using sort-by on certain fields, output is sorted unexpectedly

**What you expected to happen**: Proper numerical sort on resource request fields

**How to reproduce it (as minimally and precisely as possible)**:

On `kubectl get pods -A  -o=custom-columns='CPU:.spec.containers[*].resources.requests.cpu,MEM:.spec.containers[*].resources.requests.memory,NAME:.metadata.name' --sort-by='.spec.containers[0].resources.requests.cpu'`

Following output is returned (cropped massively):

```<none>           <none>            some-pod
2                8Gi               some-pod
<none>           <none>            some-pod
<none>           <none>            some-pod
<none>           <none>            some-pod
2                8Gi               some-pod
1                6Gi               some-pod
1                6Gi               some-pod
1                6Gi               some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
10m              <none>            some-pod
50m              100Mi             some-pod
50m              32Mi              some-pod
50m              32Mi              some-pod
```

Expected behaviour was to show larger requests (2 and 1 full vCPU) below 10 and 50 milicpus.
Also, pods without requests are just randomly thrown around at the beginning of the output.

**Anything else we need to know?**:

**Environment**:
- Kubernetes client and server versions (use `kubectl version`): 
Also on server version 19.
```Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"archive", BuildDate:"2020-11-25T13:19:56Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.12-eks-7684af", GitCommit:"7684af4ac41370dd109ac13817023cb8063e3d45", GitTreeState:"clean", BuildDate:"2020-10-20T22:57:40Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Arch Linux

### PR DESCRIPTION
**Pull request by lauchokyip**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When using `sort-by` to sort Resources, output is sorted unexpectedly

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1019

#### Does this PR introduce a user-facing change?

```release-note
Support correct sorting for cpu, memory, storage, ephemeral-storage, hugepages, and attachable-volumes
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:445]: Just a NIT: As an effort to maybe make this tests more 'readable' (and also less lines of codes!) I'm wondering if maybe it isn't a good idea to instead of keep repeating the 

```
Spec: corev1.PodSpec [....]
```

we couldn't create a helper function here, like createPodSpecResource(memreq, memlim, cpureq, cpulim).This would return the correct spec, and could be called on every part of the code that uses this. We can reduce 11 lines per corev1.Pod spec here (I've counted 242 lines), and would add...maybe 30 lines on the helper function? :)

**Review by rikatz — CHANGES_REQUESTED** [review]
@lauchokyip thanks for your PR.

I've left some comments, PTAL :)

**Comment by lauchokyip** [comment]
Thank You @rikatz, your feedbacks are very helpful!

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:445]: Great call!

**Comment by rikatz** [comment]
/lgtm

Thank you!

**Review by rikatz — APPROVED** [review]

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: Can you please help me understand why we need to do this?

Comment (eddiezane) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: Thanks for the explanation. We should avoid this dance if possible. Might need to change how we're doing the reflection. Need someone that knows Go way better than I to take a look.

@KnVerey @seans3 can either of you please take a look?

**Comment by eddiezane** [comment]
/hold

While we sort out the above

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go
Comment (eddiezane) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: Specific failure without this:

```
    --- FAIL: TestSortingPrinter/pod-resources-cpu-random-order-with-missing-fields (0.00s)
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
        panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

goroutine 78 [running]:
testing.tRunner.func1.2(0x5521bc0, 0x5a08620)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc000783200)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1146 +0x4b6
panic(0x5521bc0, 0x5a08620)
        /usr/local/Cellar/go/1.16.3/libexec/src/runtime/panic.go:965 +0x1b9
reflect.valueInterface(0x5721c80, 0xc00070a280, 0xb9, 0xc0006a3201, 0x58ca738, 0xc0006b77a8)
        /usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:1027 +0x169
reflect.Value.Interface(...)
        /usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:1016
k8s.io/kubectl/pkg/cmd/get.isLess(0x5721c80, 0xc00070a280, 0xb9, 0x5721c80, 0xc00070a2c0, 0xb9, 0xb9, 0x6f23108, 0x18)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:200 +0x15dd
k8s.io/kubectl/pkg/cmd/get.isLess(0x5728f40, 0xc00070a280, 0x99, 0x5728f40, 0xc00070a2c0, 0x99, 0x0, 0x0, 0xc0006a30b0)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:207 +0x1750
k8s.io/kubectl/pkg/cmd/get.(*RuntimeSort).Less(0xc00070a180, 0x1, 0x0, 0x30)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:341 +0x55a
sort.insertionSort(0x5a58530, 0xc00070a180, 0x0, 0x4)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:40 +0xc4
sort.quickSort(0x5a58530, 0xc00070a180, 0x0, 0x4, 0x6)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:222 +0x205
sort.Sort(0x5a58530, 0xc00070a180)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:231 +0x76
k8s.io/kubectl/pkg/cmd/get.SortObjects(0xef2c918, 0xc0005b6e60, 0xc00070a040, 0x4, 0x4, 0x582f039, 0x2b, 0xc0005b6e60, 0xc0001e9110, 0x5a25400)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:146 +0x45d
k8s.io/kubectl/pkg/cmd/get.(*SortingPrinter).sortObj(0xc00008af10, 0x5a2e680, 0xc0000e08c0, 0xc00051d680, 0x3)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:83 +0xb9
k8s.io/kubectl/pkg/cmd/get.TestSortingPrinter.func3(0xc000783200)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:601 +0xf9
testing.tRunner(0xc000783200, 0xc000406b40)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1238 +0x2b3
FAIL    k8s.io/kubectl/pkg/cmd/get      0.615s
FAIL
```

Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: I agree with @eddiezane - we should not need to do that. It seems to me that what's happening is that the test data is very differently typed from the input the command would actual deal with. Specifically, in this case the test passes in corev1.Pod objects, whereas the "real thing" deals in unstructured.Unstructured. So in the test code path, the jsonpath `.spec.containers[0].resources.requests.cpu` resolves to a `resource.Quantity` struct, so we hit [this case](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L183). We then [iterate over the fields](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L191-L196) looking at their types, which blows up because they're (mostly) private. 

So I think the correct way to fix this _for the tests_ (and any other callers that might pass in the real typed structs, since this package is public) is to add `resource.Quantity` handling right there, the same way the code already does for [`metav1.Time`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L186-L189):

```go
		if iQuantity, ok := in.(resource.Quantity); ok {
			jQuantity := j.Interface().(resource.Quantity)
			return iQuantity.Cmp(jQuantity) < 0, nil
		}
```

We probably want test coverage for the "real" code path too though! We actually kinda already have some--remove the string-type changes and the table tests fail, because their setup converts the input to unstructured.Unstructured. That said, we could explicitly write tests this way too, e.g.:

```go
func createUnstructuredPodResource(memReq, memLimit, cpuReq, cpuLimit string) unstructured.Unstructured {
	pod := &corev1.Pod{
		Spec: createPodSpecResource(memReq, memLimit, cpuReq, cpuLimit),
	}
	return *toUnstructuredOrDie(encodeOrDie(pod))
}

{
			name: "pod-unstructured-resources-cpu-random-order-with-missing-fields",
			obj: &unstructured.UnstructuredList{
				Object: map[string]interface{}{
					"kind":       "List",
					"apiVersion": "v1",
				},
				Items: []unstructured.Unstructured{
					createUnstructuredPodResource("", "", "0.5", ""),
					createUnstructuredPodResource("", "", "10", ""),
					createUnstructuredPodResource("", "", "100m", ""),
					createUnstructuredPodResource("", "", "", ""),
				},
			},
			sort: &unstructured.UnstructuredList{
				Object: map[string]interface{}{
					"kind":       "List",
					"apiVersion": "v1",
				},
				Items: []unstructured.Unstructured{
					createUnstructuredPodResource("", "", "", ""),
					createUnstructuredPodResource("", "", "100m", ""),
					createUnstructuredPodResource("", "", "0.5", ""),
					createUnstructuredPodResource("", "", "10", ""),
				},
			},
			field: "{.spec.containers[].resources.requests.cpu}",
		},
```

Wdyt? Would that be overkill?

Regardless, **I think we should have integration test coverage for this change**. Luckily, it seems to be pretty simple to add:

1. Add resource fields to [these test pods](https://github.com/kubernetes/kubernetes/tree/master/hack/testdata/sorted-pods)
2. Add a new assertion [somewhere around here](https://github.com/kubernetes/kubernetes/blob/master/test/cmd/get.sh#L317) e.g.:

```bash
 # ensure sorting by resource quantity works
  output_message=$(kubectl get pods --sort-by="{.spec.containers[0].resources.requests.cpu}")
  kube::test::if_sort_by_has_correct_order "${output_message}" "sorted-pod3:sorted-pod1:sorted-pod2:"
```

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:282]: Is it actually possible to hit both this case and the one at L185? I'm new to this code, but since the target is always structured the same, I'd expect to always hit the same case (well, one consistent case for a unstructured.Unstructured resource input and a second consistent case for a proper type containing resource.Quantity, as discussed in the other thread). Experimentation suggests to me that this is the one hit by the Unstructured case.

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: will take a deeper look soon

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: @KnVerey I actually thought of adding this, 
```
if iQuantity, ok := in.(resource.Quantity); ok {
			jQuantity := j.Interface().(resource.Quantity)
			return iQuantity.Cmp(jQuantity) < 0, nil
		}
```
but it won't reflect what happens when we want to sort a `Resource` because the `Resource` will be treated as `string` instead of `Resource` as explained in `kubectl explain pod.spec.containers.resources`
```
KIND:     Pod
VERSION:  v1

RESOURCE: resources <Object>

DESCRIPTION:
     Compute Resources required by this container. Cannot be updated. More info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

     ResourceRequirements describes the compute resource requirements.

FIELDS:
   limits	<map[string]string>
     Limits describes the maximum amount of compute resources allowed. More
     info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

   requests	<map[string]string>
     Requests describes the minimum amount of compute resources required. If
     Requests is omitted for a container, it defaults to Limits if that is
     explicitly specified, otherwise to an implementation-defined value. More
     info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

```
so I don't think changing to `unstructured.Unstructured` type is overkill

Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: I agree that the `resource.Quantity` case would not get hit when invoking `kubectl`, which is the case where the input will end up being an `unstructured.Unstructued`. The reason we might want to add that case anyway is that this code can be used in other contexts and can in fact accept specifically typed input, as your initial tests cases showed. If we don't add the `resource.Quantity` handling (similar to the Time handling already present), sorting will continue to be incorrect when the code is called directly with e.g. a `v1.Pod`. Perhaps an approver can clarify whether or not that matters.

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:337]: Sure I will keep that then 👍🏼

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:282]: I was not sure if it will hit the case `reflect.String` so I added that in just in case, will remove that 😄 . I am not sure what is annotation, can you clarify that?

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:282]: You are right, it's not needed 👍🏼

**Comment by lauchokyip** [comment]
Thank you @eddiezane  and @KnVerey for your thoughtful reviews! Let me know if anything else needs to be changed, thank you

File: staging/src/k8s.io/kubectl/pkg/cmd/get/get_test.go
Comment (lauchokyip) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/get_test.go:863]: Saw this PR and fixed some spelling mistake since I am doing the `get.go`

https://github.com/kubernetes/kubernetes/pull/101996

File: staging/src/k8s.io/kubectl/pkg/cmd/get/get_test.go
Comment (lauchokyip) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/get_test.go:902]: same as above

**Comment by eddiezane** [comment]
Great job on this one! I think we all learned a ton 😅 .

/approve

For lgtm
/cc @KnVerey @rikatz

**Comment by pacoxu** [comment]
Will this support sort of `ephemeral-storage`?

**Comment by lauchokyip** [comment]
Hi @pacoxu I believe it will but I will add more test cases tomorrow. Thank you for raising that question

**Comment by pacoxu** [comment]
This change LGTM.
I think you can fix others in a new PR.

**Comment by KnVerey** [comment]
> Do you think I should add more test cases for all of them

No, I don't think it's necessary to add coverage for additional `container.resources` keys. I agree with you that the implementation is not sensitive to the quantity being at one key in `resources` vs another, and you already have coverage for two different ones.

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:55]: Make sense!

**Comment by lauchokyip** [comment]
@KnVerey , fixed, PTAL 👍🏼

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:52]: This could just `return resource.ParseQuantity(str)`... so let's get rid of this function and inline the `resource.ParseQuantity(str)` calls.

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:100]: Any reason not to add `t.Helper()` here too?

**Review by KnVerey — COMMENTED** [review]
Just a couple more little test quibbles, but after that I think this looks ready to go!

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:100]: I just thought because that function doesnn't have `t.Errorf()` so it might not be needed 😕  will add them in

**Comment by lauchokyip** [comment]
@KnVerey fixed! ready for final look

File: staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:97]: [What `t.Helper()` does ](https://golang.org/pkg/testing/#B.Helper) is let Go know that this is a test helper, so that it can skip it when printing line and file information, which makes error reporting easier to understand when a test fails. A pretty minor thing, but since the Testing instance already has to be passed in, might as well get this benefit. 🙂

**Pull request merged**
